scabha/schema_utils.py

```diff
@@ -125,15 +125,35 @@ def nested_schema_to_dataclass(nested: Dict[str, Dict], class_name: str, bases=(
 
 _atomic_types = dict(bool=bool, str=str, int=int, float=float)
 
-def _validate_list(text: str, element_type, schema):
+def _validate_list(text: str, element_type, schema, sep=",", brackets=True):
     if not text:
         return schema.default
-    if text == "[]":
-        return []
-    if text[0] == "[" and text[-1] == "]":
+    if brackets:
+        if text == "[]":
+            return []
+        if text[0] != "[" or text[-1] != "]":
+            raise click.BadParameter(f"can't convert to '{schema.dtype}', missing '[]' brackets")
         text = text[1:-1]
     try:
-        return [element_type(x) for x in text.split(",")]
+        return list(element_type(x) for x in text.split(sep))
+    except ValueError:
+        raise click.BadParameter(f"can't convert to '{schema.dtype}'")
+
+def _validate_tuple(text: str, element_types, schema, sep=",", brackets=True):
+    if not text:
+        return schema.default
+    print(text, element_types)
+    if brackets:
+        if text == "[]":
+            return []
+        if text[0] != "[" or text[-1] != "]":
+            raise click.BadParameter(f"can't convert to '{schema.dtype}', missing '[]' brackets")
+        text = text[1:-1]
+    elems = text.split(sep)
+    if len(elems) != len(element_types):
```

Comment:
I have not checked, but aren't there two cases for tuples? `Tuple[Type1, Type2, ...]` and `Tuple[Type]` - does this check work for both?

Reply:
Good question. [From my reading of this](https://docs.python.org/3/library/typing.html#annotating-tuples), ``Tuple[Type]`` means a single-element tuple, so this check should sill be valid.

As I just learned, Python recognizes ``Tuple[Type,...]`` as representing a variable-length tuple with a typed element. But for the purposes of specifying a CLI (both when passing arguments, and when parsing arguments), this seems to me to be indistinguishable from ``List[Type]``, so I'm content to not support the former.

Reply:
Ah yeah - thanks for understanding what I meant rather than what I typed. That makes sense to me!